calyx/opt/src/passes_experimental/fsm_builder.rs

```diff
@@ -0,0 +1,276 @@
+use crate::analysis::{IncompleteTransition, StaticSchedule};
+use crate::traversal::{Action, ConstructVisitor, Named, Visitor};
+use calyx_ir::{self as ir, GetAttributes};
+use calyx_utils::CalyxResult;
+const ACYCLIC: ir::Attribute =
+    ir::Attribute::Internal(ir::InternalAttr::ACYCLIC);
+
+pub struct FSMBuilder {}
+
+impl Named for FSMBuilder {
+    fn name() -> &'static str {
+        "fsm-builder"
+    }
+    fn description() -> &'static str {
+        "generates medium fsms in one pass for static and dynamic"
+    }
+}
+
+impl ConstructVisitor for FSMBuilder {
+    fn from(_ctx: &ir::Context) -> CalyxResult<Self> {
+        Ok(FSMBuilder {})
+    }
+    fn clear_data(&mut self) {}
+}
+
+// node that passes into this function is a part of the schedule
+impl StaticSchedule<'_, '_> {
+    fn build_abstract(
```

Comment:
Since there are many arguments and several return values here, a longish doc comment describing what this function does might be helpful!